### ISSUE TITLE
Error: The request was aborted because there was no available instance.

### ISSUE THREAD

**Issue by dimavedenyapin** [firebase/firebase-functions#962]
This is happening in production firebase environment with a Blaze subscription, I've started seeing the error `The request was aborted because there was no available instance.` since 22nd August 10pm GMT+8.
This error happens across all functions when I make 100+ invocations. 
When the error appears it affects all other functions as well (see screenshot).
Can happen with any function with `maxInstances` parameter set or without it.

All functions are deployed in `us-central1`
Quotas doesn't seem to reach the limit.

### [REQUIRED] Version info
**node:**
v12.22.3

**firebase-functions:**
3.14.1

**firebase-tools:**
9.16.0

**firebase-admin:**
9.11.0

### [REQUIRED] Test case
Firebase pubsub listener

### [REQUIRED] Steps to reproduce
Send 100+ messages to the firebase pubsub

### [REQUIRED] Expected behavior
Functions execute.

### [REQUIRED] Actual behavior
Functions failing with a message: `The request was aborted because there was no available instance.`
![image](https://user-images.githubusercontent.com/14056855/130712795-75646748-2076-4442-9a66-1cbd955c1d3e.png)

### Were you able to successfully deploy your functions?
successfully deployed

**Comment by taeold** [comment]
Thanks for reporting the issue here. I noticed that there were several GCP support ticket opened w.r.t. to this issue, and will post any relevant updates here for wider audience as the GCF team makes progress on the issue.

**Comment by dvrfluxchat** [comment]
Its happening in production
We're also experiencing this since August 23rd running on Node 12 instances on asia-south1.

It doesn't have any correlation to the load, I run a hourly pub sub function which also fails frequently. Critical messages are being dropped because of this.

**Comment by michael-ottink** [comment]
![image](https://user-images.githubusercontent.com/15832303/131353366-c10c0502-46fa-47d6-a83b-fccfeda93feb.png)

We are also getting this error on europeWest1 There are just 2 people testing the application right now and this already happens. In about a week it will need to scale to thousands of people, is there anything I can do about this?

**Comment by stoy** [comment]
Same error for me. Functions randomly return  **The request was aborted because there was no available instance.** .
Until now, never had this error.

**Comment by wneild** [comment]
Same issue since ~ 2021-08-26 20:00 BST,

Node 14,
firebase-functions 3.14.1
firebase-admin 9.10.0

It is happening with _very_ minor spikes of requests < 100 across all function deployments.

I'd also like to understand why this very impactful issue being reported by many people also isn't reflected on https://status.cloud.google.com/ as being investigated.

**Comment by robin-whg** [comment]
Same here. The requests take way too long. A simple function with a single Firestore write takes up to 6000ms if it isn't aborted.  Europe West 1

**Comment by dimmetrius** [comment]
Same issue since 2021-08-23

**Comment by damienromito** [comment]
Same issue, since this morning, and it's getting worse !!!
I tried to 
- deploy each function again ❌
- delete then add function ❌

I sent a bug report to Firebase...

**Comment by tfang17** [comment]
Same issue.

**Comment by taeold** [comment]
Hi everyone.

Google Cloud Function (GCF) users as a whole are reporting the same issue described here, and https://issuetracker.google.com/issues/153207649#comment3 is the official response from the GCF team.

tl;dr GCF nodejs runtime used to silently drop requests when instance couldn't be scaled fast enough to respond to demand. Now it's logging the failed request on your project's log, hence the sudden appearance of the issue ([release note](https://cloud.google.com/functions/docs/release-notes#July_07_2021)). For pubsub-triggered functions, this error is usually handled gracefully by automatic retry mechanism in the GCF infrastructure. The same can't be said of HTTP-triggered functions, and the request would have been dropped by the client unless a retry mechanism was already implemented.

To reduce occurrence of the once invisible but now transparent "aborted because there was no available instance" errors, recommendations in https://cloud.google.com/functions/docs/troubleshooting#scalability applies.

I hope this clears up the confusion a bit. I'll leave this ticket open to answer any follow up questions, but since this problem is directly related to Google Cloud Functions and not specific to Firebase Functions, please consider reaching out to [GCP support](https://cloud.google.com/functions/docs/getting-support) with your project-specific questions.

**Comment by larssn** [comment]
@taeold But could you clarify what happens to event-driven functions, such as firebase/firestore triggers? The docs [here](https://cloud.google.com/functions/docs/bestpractices/retries#semantics_of_retry) guarantees at-least-once execution. 

I'm hoping that is still the case. Only few of our functions are idempotent enough to warrant enabling the retry policy.

**Comment by taeold** [comment]
@larssn At-least once guarantee applies to all event-driven functions. Are you seeing events from Firebase/Firestore being dropped on your project?

**Comment by larssn** [comment]
> @larssn At-least once guarantee applies to all event-driven functions. Are you seeing events from Firebase/Firestore being dropped on your project?

It's not possible to see what data exactly is affected, due to the nature of the error message. I'm just hoping that our triggers are retried.

If so, then I'm wondering why it's necessary to show it as an error, couldn't an "info"-level message suffice?

**Comment by taeold** [comment]
@larssn Same thoughts on the error message. I think this is first step on exposing the log to GCF users, and the team is working on a fix to output the log as a warning if it is going to be retried.

**Comment by dgobaud** [comment]
We are also seeing this https://github.com/firebase/firebase-functions/issues/965 started US central August 24

it was flooding our logs we had to exclude all such messages from alerts

google Case 28781796

**Comment by dimavedenyapin** [comment]
@taeold thanks for your research and updates 👍
I am glad to hear that it still guarantees the At-least once guarantee applies to all event-driven functions.

Looking forward to a fix the log outputs.

**Comment by npicouet** [comment]
I don't really understand how its only a silent vs logging problem.
Our app has been running without issue for months, and only NOW are we getting these issues that actually are preventing some functions from running. It feels to me like something actually is going on or has changed, because we have not changed our backend for a while and its been running smoothly until now.

**Comment by deepak786** [comment]
Earlier there were timeout errors but now we are getting these errors.
![Screen Shot 2021-09-03 at 9 54 26 PM](https://user-images.githubusercontent.com/14223997/132038211-fbeec0cb-75c2-4232-8649-7b020bbbe59f.png)

**Comment by taeold** [comment]
@npicouet @deepak786 Are these coming from HTTP functions or event-triggered functions? I've been told that these errors are "okay" for event-triggered functions in the sense that Google Cloud Infrastructure will retry the failed functions invocations. So your events aren't actually being dropped and your application is still guaranteed at-least once delivery of messages.

I'm sounding like a parrot, and I'm sorry for doing this over and over again - if you are seeing serious issues, please go ahead and contact [Google Cloud Support](https://cloud.google.com/functions/docs/getting-support). I'm not aware of any outage that explains the issues being reported here.

**Comment by deepak786** [comment]
> Earlier there were timeout errors but now we are getting these errors.
> ![Screen Shot 2021-09-03 at 9 54 26 PM](https://user-images.githubusercontent.com/14223997/132038211-fbeec0cb-75c2-4232-8649-7b020bbbe59f.png)

In this screenshot, all the functions are event-triggered functions.

**Comment by pedronieto84** [comment]
In my case the function that has that error message is a function triggered by a "firestore trigger" ".onCreate".

**Comment by Gbuomprisco** [comment]
This started happening today to me, HTTP Functions

**Comment by markgoho** [comment]
Adding in a log that shows the instance not available, but the retries and eventual successful run of a scheduled cloud function: 
![image](https://user-images.githubusercontent.com/9759954/132234622-f5f6be95-cad6-4b42-8069-25376aa9b410.png)

Is everyone _only_ seeing the failures with no successful run?

**Comment by michael-ottink** [comment]
I have a http function and when I litterally call it twice at the same time, it already shows this error. This happens every time. This could never work for thousands of users.

**Comment by petermiles** [comment]
Started happening to us yesterday for HTTP functions as well. Across our environments, this is happening quite often.

**Comment by sanekyy** [comment]
save issue:(

**Comment by ChrisForeman** [comment]
Started at 4:30PM EST for me. I'll be in hot water if this persists 😬

**Comment by hatboysam** [comment]
I am also experiencing this issue.

**Comment by taeold** [comment]
Friendly reminder - please reach out to [Google Cloud Support](https://cloud.google.com/functions/docs/getting-support) if your **HTTPS** (not event triggered) functions are having problems.

We are keeping this issue open make it easy for users experiencing the issue described in the original post to understand what's going on and to make the right next steps. Unfortunately, the issue being reported here cannot be fixed by code changes in this repository.

**Comment by gbourne1** [comment]
Same here - multiple functions keep getting: "The request was aborted because there was no available instance"

**Comment by Hivemind9000** [comment]
Happening to me too in the past 24 hours or so. I am monitoring uptime via Honeybadger and it has only just started detecting this issue (via HTTP request). Almost all of my cloud functions are failing, and most are non-event triggered (http onRequest or onCall).

**Comment by sebagutierrez** [comment]
We've had this issue for a while now. It seems to be happening more often lately.

**Comment by TrustyTechSG** [comment]
Im having the same issue here. Start about 1 weeks ago, never seen this error in the pass 1 year of using firebase functions.

**Comment by sarimabbas** [comment]
For me it started about a week ago as well. My code has not been updated since July, so it is definitely a new problem

**Comment by mnahta** [comment]
Only started happening few hours earlier on cloud functions. Re deployed but doesn't seem to be working.

**Comment by ToeFungi** [comment]
Our cloud functions have been able to scale with no issue for over a year. Nothing has changed within our infrastructure but as of late last night, our functions have begun to fail. This doesn't appear to be related to traffic, cold starts or long running executions. A request will be made to a function and it will fail on every request for several minutes. It will then begin to work and another function will begin to fail.

There definitely seems to be something larger going on here than just revealing logs to Stackdriver.

**Comment by sanketplus** [comment]
My theory is that they reduced tolerance for cold starts. For example: earlier they were ok waiting 2s for cold starting but now they throw the said error in just 1s. If you notice (or can create) a function with lil to no dependency, basically a helloWorld function, will not get affected by this.

This announcement also aligns with [announcement of min-instance](https://firebase.google.com/docs/functions/manage-functions#min-max-instances) for cloud function. Support recommended using this new beta feature but did not have an answer for the cause of this issue.

So likely they changed some configuration in the backend and this is an effect of that. Lot more people having production impact because of this here: https://issuetracker.google.com/issues/194948300 Hope they find and fix this soon 🤞

**Comment by charlierushton** [comment]
Happening to me too. No issues with scaling since we started our project in March 2020 until start of last week, when this issue happens every few minutes

**Comment by Wtrapp** [comment]
Same issue for us. We haven't made any changes to cloud functions.

**Comment by taeold** [comment]
Hot off the press - Google Cloud Functions in us-central1 did report some problem ~2021-09-08:

https://status.cloud.google.com/incidents/16SSwVXrYSLjy8fEMvyZ

The status report claims that the issue only affected functions deployed in us-central1 and that it is now resolved. If you are still seeing issues, please contact [Google Cloud Support](https://cloud.google.com/functions/docs/getting-support).

**Comment by tolypash** [comment]
I am getting the same issue. 1 year using cloud functions no problem (europe-west3)

Now this is happening out of nowhere for the past 3 weeks. Contacted firebase support, they tried to tell me to reproduce this problem, however it is impossible to reproduce on demand because it is so random. I explain to them that this happens for all functions, for a few minutes each time (as if it's an outage). Not sure what to do right now :/

**Comment by sanketplus** [comment]
@tolypash not exactly the solution you'd like but consider moving out of google's ecosystem. This is a lesson learned hard. So far I have only heard about poor customer support but now I have witnessed it with this issue.

**Comment by larssn** [comment]
@sanketplus This is not the place for that.

This thread is dedicated to figuring out if there's a problem, and how to resolve it.

**Comment by sanketplus** [comment]
@larssn I get what you're saying. OP is saying support is not being helpful and so was the case with me when I was helping someone navigating this issue. I think suggestion/solution of moving out of this is pragmatic. More so when you are having a real customer impact which is making you lose money. You do not want to bet your company and its revenue to a cloud company which is having hard time determining if at all there is a problem.

Anyway, that is my personal take on this. You are welcome to disagree with it :)

**Comment by Hivemind9000** [comment]
@tolypash The problem seemed to go away for 4-5 days (after Google said they found/fixed the issue causing the high frequency of errors) but is now back this morning. That said I've only had 5 errors this morning from approx 5000 calls, and nothing for the past 6 hours (from another 5000 calls) so not as frequent/severe as before. It may have been a transient load issue in the US central region.

I think it might be helpful to read this thread from the Google team:

https://issuetracker.google.com/issues/194948300#comment24  (posted today)
https://issuetracker.google.com/issues/194948300
and https://cloud.google.com/functions/docs/concepts/exec#execution_guarantees

My reading is that these errors were occasionally occurring before (primarily due to load spiking), but not being thrown as an error. As noted in the issue tracker thread, event driven functions will retry on error, while http ones don't - it's up to your client to perform retries on errors. At the current error rate I am experiencing I think the retry strategy is reasonable (necessary even since Google state they don't guarantee HTTP executions, and the Internet in general being a relatively unreliable network).

I guess it depends on how many errors you are seeing, and the error ratio compared to successful GCF executions?

**Comment by u007** [comment]
im having the same issue on asia-northeast1

**Comment by YUTOPASO** [comment]
I am getting the same issue on asia-northeast1.

All functions failed between 15:00 and 16:00 today.

**Comment by YongjinK** [comment]
@taeold You are lying. This issue is not about invisible warning. Our game service has been calling 100,000 https calls without error for 1 month. We are just experiencing the same issue reported here since yesterday (asia-northeast1). We had to handle 6+ refund issues today because of this issue. Be honest, Google. Tell us what you guys are doing??

**Comment by dgobaud** [comment]
We used to have a similar problem using Firebase for HTTP serving - not errors but cold starts causing HTTP requests to take 10+ seconds meaning our app would often hang loading looking like it crashed. Seems a change has turned what were cold starts into errors.

The problem with Firebase is there is no way to control cold starts unlike with AWS Lambda. Lambda is much smarter about scaling up and down and sending requests to existing instances whereas with Firebase it is more random. Eg having a pinger keeping an instance alive doesn't really do anything useful.

The solution is to stop using Firebase for HTTP... it really is very bad for it. Switch to App Engine and you can control the scaling a lot more and avoid these problems.

**Comment by sanketplus** [comment]
+1 @dgobaud we also tried to keep the functions warm, not helping anymore. Similar to @YongjinK's experience, the failure started happening suddenly with no change from our side. There is definitely something that was changed from google's side and they are failing to ack it and work on it.

Though a few individual think that this is not a solution, I +1 on stop using GCloud functions, more so when it is resulting into prolonged customer impact.

**Comment by charlierushton** [comment]
Hi all,

I last saw this issue about a week ago and posted in this thread, and I haven't seen it since. 

All I did was edit the function in Google Cloud Console (not Firebase Console) and set the minimum instance to 1, rather than leaving min & max to empty, see below screenshot.

After setting this, I haven't seen this error in the functions log. Could be coincidence, I don't know, but this seems to have worked for me. I will post again if it occurs again. 

<img width="544" alt="Screenshot 2021-09-14 at 16 07 34" src="https://user-images.githubusercontent.com/74256062/133283577-63058c0d-fbed-45a5-bf63-21a905f78a85.png">

**Comment by charlierushton** [comment]
Apologies, my last post is not true. I have just filtered my log to see that other user's of my cloud functions are encountering this error, but not as often as when I first posted.

<img width="1027" alt="Screenshot 2021-09-14 at 16 12 01" src="https://user-images.githubusercontent.com/74256062/133284320-06625c01-aedb-4639-bb15-8bad07c60ef3.png">

**Comment by sanketplus** [comment]
@charlierushton that solution will work indeed. But then you start paying north of 6 USD **per instance**. And this is pre-GA feature, so no stabilities are guaranteed.

This is what customer support suggested to use. which beats the purpose of having serverless functions and pay for only what you use.

src: https://firebase.google.com/docs/functions/manage-functions#min-max-instances
> Note: A minimum number of instances kept running incur billing costs at idle rates. Typically, to keep one idle function instance warm costs less than $6.00 a month. The Firebase CLI provides a cost estimate at deployment time for functions with reserved minimum instances. Refer to Cloud Functions Pricing to calculate costs.

**Comment by dgobaud** [comment]
It seems minInstances is a new thing to help with the cold start problem - I'm guessing still when scaling up something has changed so cold starts cause errors instead of just very slow returns

**Comment by gbourne1** [comment]
It looks like the issue is back - dozens of "The request was aborted because there was no available instance."

**Comment by Wtrapp** [comment]
Yep. This issue is back for us too on us-central

**Comment by hatboysam** [comment]
I’m getting dozens of these errors again for HTTP functions, it seems even
worse than before.

**Comment by charlierushton** [comment]
@sanketplus you are correct. checked my billing accout this morning to find 20GBP billed for setting min instances to 1 for 7 days. Set min instances to zero very quickly.

**Comment by sceee** [comment]
Can also confirm, last night this error occurred multiple times again on a production function running in us-central1. 

The days before that (from last Sunday), those errors did not occur.

**Comment by akshatflx** [comment]
Have been seeing a lot of such errors in `us-central1` since yesterday evening (Sep-14, GMT).

**Comment by charlierushton** [comment]
I can confirm after setting min instances to zero, the issue has returned, seeing a lot of errors in my logs. also receiving a lot of phone calls & support tickets from customers reporting issues they have not had before. 

we are losing money processing refunds and spending time on customer support when this issue is above our head

customers reported time of the issue lines up with the time in the logs when the error occurs. surely must be an answer from google sometime soon?

**Comment by taeold** [comment]
I hear your frustrations and am truly sorry for all the disruptions.

This issue is affecting all Google Cloud Functions customer, not just Firebase users, and this issue isn't a communication channel monitored and used by the Google Cloud Functions team.

Please follow https://issuetracker.google.com/issues/194948300 for latest updates, and if you have specific issues in your project you'd like to escalate please contact [GCP support](https://cloud.google.com/functions/docs/getting-support).

We are keeping this Github issue open to help redirect users to appropriate support channels. I'll do my best to escalate reports being made here, but again, [GCP support](https://cloud.google.com/functions/docs/getting-support) is the best way to get attention of "Google".

**Comment by shauryaaher** [comment]
I'm kind of intermediate at using Cloud Functions for Firebase. Even though I didn't face this issue. I'd still like to know about what to do, if I face it in the future.

**Comment by shauryaaher** [comment]
> I'm kind of intermediate at using Cloud Functions for Firebase. Even though I didn't face this issue. I'd still like to know about what to do, if I face it in the future.

BTW, my Functions are at us-central1.

**Comment by Hivemind9000** [comment]
We had a resurgence of issues yesterday on US Central - about 12 throughout the day (from 15k requests). Implementing retries for HTTP requests helped ensure that we didn't have a service disruption for our customers. Today we've only had 2 errors show up in the past 6 hours (from about 4k).

It seems that Google has been struggling intermittently over the past few weeks to keep up with server demand, and I am not confident it will go away completely anytime soon. I recommend to everyone to implement a decent retry strategy in your client (if you haven't already). As stated in the docs, the HTTP requests are not guaranteed to go through, while event driven functions will have 1 retry.

Also consider setting the minimum instances to ensure you're critical functions will not have to cold start (and therefore request a server from the pool). As mentioned above, this comes at a cost of US$6/Mth/function. Might be worth it for some (we are doing this for a few of our most latency-sensitive functions).

Some further reading:

For Pubsub / background / event-based triggered functions the requests running into this issue will be retried automatically hence there is no loss of information in terms of requests reaching GCF.

For HTTP triggered functions , the client is responsible for retrying a request ( with recommended exponential backoff + retry methods ) that ends with this error message.

References:

[1] - https://cloud.google.com/functions/docs/concepts/exec#execution_guarantees
[2] - https://cloud.google.com/functions/docs/bestpractices/retries#semantics_of_retry

**Comment by eliezerbs9** [comment]
Just now experiencing it for the first time at us-cental1.

**Comment by stevecode21** [comment]
Same issue here.

**Comment by iocuydi** [comment]
Same issue, just experienced on US central 1 on two different projects within the past hour

**Comment by antoniooi** [comment]
Please fix your Cloud Function's instance scalability issue ASAP. We're not going to pay for your error for US$6/Mth/Function just for the workaround. Please do not get the users to responsible for your failure to fulfill your GCP's "high scalability" commitment. If causing problem will lead to more profits, will there be a necessity to solve the problem? Stop asking the users to pay for the workaround solutions and stop saying it is "worth for the money". Just get it done and don't let people losing hope on Google. Thank you.

**Comment by rajivsubra1981** [comment]
Same issue faced by us on us-central1 - I wonder if the workaround of $6/month/function is actually making the problem worse, with folks buying up instances they wouldn't actually need if there was availability now exacerbating the shortage of instances.

**Comment by Hivemind9000** [comment]
Everyone, if you can please post that you are getting the errors on this thread here:

https://issuetracker.google.com/u/1/issues/199180393

More posts might hopefully get Google's attention...

**Comment by samodadela** [comment]
We are also seeing similar problems. Java cloud functions that are triggered by PubSub are logging warnings but in the end all messages were processed. BTW: the warnings don't stop even when max instances (in our case 10) is reached. 
![image](https://user-images.githubusercontent.com/16334222/133572637-6b15a47b-5c9f-4f4e-bf04-0bcd1d771d79.png)
What's the purpose of the warning anyway? What can I do to get rid of it (apart from setting minInstances - which is not desirable)?

The **more serious problem** is that CFs triggered by **Cloud Scheduler just error** and **do not trigger the cloud function**. Our imports are triggered by a cron job every day. Before this we had 0 problems with the job being triggered - it never failed to start. This week the jobs failed for 2 consecutive days. 

Please check your statement that this is not causing problems. The number of posts here is an indicator that there is something fishy going on

**Comment by crrobinson14** [comment]
This thread doesn't seem like it needs more "venting," but the linked issue-tracker posts don't allow commenting so I guess I have no choice but to chime in here.

As I see it, from a developer's perspective the problem here is twofold:

1. Yes, retrying requests is generally a Good Thing. But despite some side references above, this "need" is very lightly documented in the official Firebase docs. There are lots of claims about how it infinitely scales, but nothing I can find about how it might fail. Please tell me what I'm missing from this doc:
https://firebase.google.com/docs/functions/callable
None of the docs include "exponential backoff / retry" wrapper code or any hint that this is going to be a regular problem, and it's certainly not a problem with other FaaS offerings. It's hard to avoid the impression that Google is shifting blame for a genuine problem.

2. The whole "no change was made, all that's happened is we're logging something you never saw before" claim doesn't wash. As others have reported above, our app is now (last 48 hours, that's why I'm here) suddenly experiencing a dramatic increase in this type of failure. But it's not related to a workload spike. The function our app is calling is in development. It only gets called once an hour at most, by our development team. We're seeing about a 5% failure rate when we do, but it's certainly not workload related. 5% doesn't sound like a lot but we're lucky we caught it in dev, because if we deployed this now we'd be swimming in user complaints.

Honestly I think @samodadela's https://github.com/firebase/firebase-functions/issues/962#issuecomment-920677241 above is my biggest fear. Sure, we can go re-engineer our app to add exponential backoffs to all calls. But Firebase has no mechanism to do that for scheduled functions or firestore triggers. "Just add retry options" CAN'T be the final answer here.

**Comment by dgobaud** [comment]
> Honestly I think @samodadela's [#962 (comment)](https://github.com/firebase/firebase-functions/issues/962#issuecomment-920677241) above is my biggest fear. Sure, we can go re-engineer our app to add exponential backoffs to all calls. But Firebase has no mechanism to do that for scheduled functions or firestore triggers. "Just add retry options" CAN'T be the final answer here.

luckily scheduled functions and triggers are guaranteed at least once working deliver it seems

**Comment by sceee** [comment]
I also second the "retrying HTTP requests is generally a good idea" but - as @crrobinson14 mentioned - based on the Firebase docs I did not realize this as something being crucial to **each** app. 

Until now, I was more in a feeling like:
of course, HTTP functions _can_ fail to execute if something really goes wrong in GCP. Therefore, they might fail in 0.000x% of all executions because of such an error. 

And for this very rarely expected case, until now I just did not handle this case in the applications code as when this happens, a user could retry the action that executes the call to the HTTP function. 

But if we're now talking about magnitudes of several percent (or even more) of executions that suddenly fail because of this error, it's a whole other level than:
"Usually the HTTP invocations work (expect some very rare outages)"

**Comment by tolypash** [comment]
I would also like to note that "retrying" these functions from the clients side is not possible either, because this issue seems to be affecting all functions for a certain period of time (ranging from a few seconds to sometimes minutes)

So even when I retry on the client, there will be another error thrown unless retries are spaced out minutes apart, which is not possible for the client.

**Comment by amitrao17** [comment]
I'm seeing this on a scheduled function that runs once an hour. it gave an error once today and once yesterday. it does appear to be automatically retried but the function should update the realtime database which is not happening. so, the function initially errors, then runs but not correctly...

**Comment by skizzo** [comment]
Same thing happening to **all** of my Firebase projects. Might as well find another job.. :(

**Comment by dgobaud** [comment]
@amitrao17  > I'm seeing this on a scheduled function that runs once an hour. it gave an error once today and once yesterday. it does appear to be automatically retried but the function should update the realtime database which is not happening. so, the function initially errors, then runs but not correctly...

what should update realtime database? we see same thing with crons error then they run correctly

**Comment by amitrao17** [comment]
The scheduled function runs every hour, does a 'fetch' to an API to get the weather, then console.log the result. Finally, it calls a '.set' to write the data into the realtime database. there are 8 writes to the realtime database which return promises and the function returns all of them.

I know the function gets called from the console logs, and i also know the database doesn't get updated. So it appears that the function sometimes never runs (with the "Error: The request..." message), sometimes starts, but doesn't let the promises go to completion.

One other thing that is happening...sometimes the function fails to write to the database the first time it is called as well -- there are multiple occasions when the function clearly makes the API call but the database does not get updated values.

I could try to 'await' all the promises before returning...

**Comment by dgobaud** [comment]
@amitrao17 i think you should definitely await all the promises before returning - I'm guessing part of what is happening is whatever Google changed is killing functions much more quickly after they finish running which per the spec is correct or at least not unexpected. Maybe before Google let functions hang around longer so for example your unresolved promises had time to finish even though there was no guarantee of that

**Comment by amitrao17** [comment]
> @amitrao17 i think you should definitely await all the promises before returning - I'm guessing part of what is happening is whatever Google changed is killing functions much more quickly after they finish running which per the spec is correct or at least not unexpected. Maybe before Google let functions hang around longer so for example your unresolved promises had time to finish even though there was no guarantee of that

Interesting. I'm new to this (a hardware guy) so my understanding was that letting the promises get finished after returning was fine. Thanks!

**Comment by rajivsubra1981** [comment]
Promises must be awaited or returned in any event to guarantee execution in
cloud functions. This has been & is the expected behavior prior to this
issue.

**Comment by amitrao17** [comment]
> Promises must be awaited or returned in any event to guarantee execution in cloud functions. This has been & is the expected behavior prior to this issue.

Yes, I was aware to return them...but it turns out that my function was not declared 'async' so that's not good...

**Comment by dgobaud** [comment]
We originally ran our client facing HTTP service on Firebase before minInstances was supported.

It was completely unusable because of cold starts. Random requests would hit the cold start and take 10+ seconds to return basically killing our client performance. With Firebase there was no way to control this - the way it scales didn't support trying to keep an instance alive so eg a pinger which works with Heroku to keep an instance alive doesn't work.

We changed to App Engine with min_instances and the problem was fixed.

It isn't clear but it seems Google did something recently to perhaps kill functions more quickly and maybe not let http requests wait for cold start and now just error. I've looked at the Google Issue Trackers to me it seems they are just saying this is a logging change but that seems wrong from what people are seeing:

1. actual http errors
2. cloud functions seemingly being killed more quickly

I can't tell from above but setting minInstances to 1 should help - people seem to report it doesn't though so unclear. If you want this fixed ASAP since it seems Google isn't going to do anything since stuff is probably technically working under spec you should first set minInstances to 1 and if that doesn't work switch to App Engine for HTTP with min_instances 1.

**Comment by DzTheRage** [comment]
Unable to comment on linked issue tracker so might as well make my voice be heard here as well. 

As other have stated in this thread, this issue is not just a logging issue and is causing actual problems for our clients. 

My best guess is there is some scalability issues that have not been truly resolved on GCP's side that are causing cloud functions to fail such as this: https://status.cloud.google.com/incidents/16SSwVXrYSLjy8fEMvyZ

**Comment by antoniooi** [comment]
Don't tell me this is just another similar approach of Google, like how they started the Cloud Functions for free via the nice packaging of Firebase, then later taking the Cloud Build and Registry Container as a reason to force everyone to move to Blaze Plan (and activate their billing), and later the US$6/Mth/Function to increase the instances or else your Cloud Functions no longer scale? So they'll say this is still considered "Pay as you Go" because your traffic is increased and you need to pay for more instances for your app to scale well? If this is the case, then this issue is more like a high level Google management issue than just a "technical issue". I hope Google isn't just making up the problem for their pricing structure to move one step forward for greater profitability.

**Comment by mbleigh** [comment]
Hi folks, we appreciate all the passionate feedback and we can tell how important this issue is to you all. The Cloud Functions team is well aware of the issue and several internal discussions and investigations are happening now. The best place to keep track of ongoing updates related to this error is here: https://issuetracker.google.com/u/1/issues/194948300

To reiterate what has been said before: there has been no intentional change in behavior to the scheduling of instances for Cloud Functions. What has changed recently is that a log line is being written when a request fails due to no instance being available. When that results in a `429` error, it's because all instances specified via `maxInstances` are saturated. When that log is a `500` error, it's because of internal unavailability of instances.

I see from many of your comments that this seems to be happening much more frequently than you expect, and on functions/instances with low traffic and no `maxInstances` set. If this is the case, please contact [Google Cloud support](https://cloud.google.com/functions/docs/getting-support) as it will require digging into the specifics of what's happening in your project to diagnose.

I'm closing this thread to further comment as there's nothing specific to the Firebase SDK for Cloud Functions that can be done to address problems related to this error message. Myself and the other members of the Firebase team will continue to follow up on this internally to make sure that all efforts are being made to get to the bottom of any regressions in instance scheduling behavior.

**Comment by mbleigh** [comment]
Hey folks, we have identified and resolved a potential cause of increased errors of this type. A sampling of customers leads us to believe that these errors may have been reduced significantly. You will still see errors of this kind where requests previously failed silently, but these errors should be rare.

If you are still experiencing this error frequently, please [reach out to support](https://cloud.google.com/functions/docs/getting-support) as outlined above.
